### DISCUSSION TITLE
What exactly is RawDeployment?

### DISCUSSION THREAD

**Discussion by iciclespider** [kubeflow/kfserving#3505] [General]
I have looked high and low in the documentation and the source code, but I have been unable to find a nice explanation of what Raw Deployment does, and does not, do. I did a quick test and it appears it does not do storage initialization, is that correct?

**Comment by MutuT2** [comment]
Kserve supports two types of deployment modes. one is Serverless mode and another one is RawDeployment mode. 

1. We can use the Serverless mode to enable autoscaling based on request volume and supports scale down to and from zero using Knative.

2. We can use the RawDeployment mode to enable InferenceService deployment with Kubernetes resources Deployment, Service, Ingress and Horizontal Pod Autoscaler. Comparing to serverless deployment it unlocks Knative limitations such as mounting multiple volumes. But "Scale down to and from Zero" is not supported in RawDeployment mode.

**Note**: Both the Serverless and RawDeployment mode does the storage initialization.

**Reply by iciclespider** [reply]
Thanks, that help. However, I was unable to get storage initialization to occur in RawDeployment mode. The storage initializer side car container was not created.

**Reply by MutuT2** [reply]
It should up and run. I hope you followed [this doc](https://kserve.github.io/website/master/admin/kubernetes_deployment/). 
Could you please share yaml file.

**Reply by iciclespider** [reply]
In a RawDeployment environment, I deployed the Getting Started model of;
```yaml
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  namespace: kserve
  name: "sklearn-iris"
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      storageUri: "gs://kfserving-examples/models/sklearn/1.0/model"
```
And the resulting Pod only has a single container:
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    internal.serving.kserve.io/storage-initializer-sourceuri: gs://kfserving-examples/models/sklearn/1.0/model
    prometheus.kserve.io/path: /metrics
    prometheus.kserve.io/port: "8080"
    serving.kserve.io/deploymentMode: RawDeployment
  creationTimestamp: "2024-03-15T17:08:33Z"
  generateName: sklearn-iris-predictor-6bf6fc48b8-
  labels:
    app: isvc.sklearn-iris-predictor
    component: predictor
    pod-template-hash: 6bf6fc48b8
    serving.kserve.io/inferenceservice: sklearn-iris
  name: sklearn-iris-predictor-6bf6fc48b8-5jhlg
  namespace: kserve
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: sklearn-iris-predictor-6bf6fc48b8
    uid: addcd915-7368-4d65-8333-8df959d034f2
  resourceVersion: "235135635"
  uid: 0b966071-b6e0-4530-905e-aa51ac06c804
spec:
  containers:
  - args:
    - --model_name=sklearn-iris
    - --model_dir=/mnt/models
    - --http_port=8080
    image: kserve/sklearnserver:v0.12.0
    imagePullPolicy: IfNotPresent
    name: kserve-container
    readinessProbe:
      failureThreshold: 3
      periodSeconds: 10
      successThreshold: 1
      tcpSocket:
        port: 8080
      timeoutSeconds: 1
    resources:
      limits:
        cpu: "1"
        memory: 2Gi
      requests:
        cpu: "1"
        memory: 2Gi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-fv28z
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: ip-10-0-185-10.ec2.internal
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-fv28z
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
```
That pod is non-functional without the storage initializer init container setup.

The organization of where this diagram https://kserve.github.io/website/master/images/controlplane.png has the "Raw Deployment" label implies that storage initialization is not part of RawDeployment, but is done at a higher level.

**Reply by iciclespider** [reply]
OTOH, I just noticed the "Serverless" label  below the "Raw Deployment" label, so I suppose that image is not implying that RawDeployment is not part of storage initialization. However, in my RawDeployment environment, that Getting Started example above did not initialize the storage.

**Reply by iciclespider** [reply]
I just found this issue that I believe is the same situation: https://github.com/kserve/kserve/issues/3499

**Reply by israel-hdez** [reply]
Would you confirm if you are using the same 0.11.0 version as in #3499?

**Reply by iciclespider** [reply]
I am using v0.12.0, from a kubectl describe pod of the kserve-controller:
```
    Image:         kserve/kserve-controller:v0.12.0
    Image ID:      docker.io/kserve/kserve-controller@sha256:59d7d878a430722a7c7ad9a95bef082fb3ddc419506bdacf3f8a9ceeb2d5975d
```

**Reply by iciclespider** [reply]
And this is the kustomization.yaml configuration I am using to deploy KServe:
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: kserve

resources:
- https://github.com/kserve/kserve/releases/download/v0.12.0/kserve.yaml

patches:
- target:
    kind: ConfigMap
    name: inferenceservice-config
  patch: |
    - path: /data/deploy
      op: replace
      value: |-
        {
            "defaultDeploymentMode": "RawDeployment"
        }
    - path: /data/ingress
      op: replace
      value: |-
        {
            "disableIstioVirtualHost": true,
            "disableIngressCreation": true,
            "ingressDomain": "svc.cluster.local",
            "ingressGateway" : "disabled",
            "ingressService" : "disabled"
        }
```

**Reply by iciclespider** [reply]
I just reread the comments in this issue https://github.com/kserve/kserve/issues/3499 and in my case I was also deploying the InferenceService to the `kserve` namespace. When I deployed it to a different namespace, storage was properly initialized. Mystery solved I believe. However, some documentation is needed clarifying this. Actually, IMO, I do not see a reason for not supporting InferenceServices in the kserve namespace.

**Reply by iciclespider** [reply]
On the other, other hand...

When deployed to the "kserve" namespace in RawDeployment mode, the kserve operator did create the Deployment and the Pod did get created. It was then the kserve mutating webhook that did not process the needed storage initialization. If deploying to the kserve namespace is not supported, the kserve operator should not be creating the Deployment and other resources. If it is supported, storage initialization is broken.
